### ISSUE TITLE
freeCodeCamp's JS Curriculum, How Can It Be Improved?

### ISSUE THREAD

**Issue by FrontEndWebDeveloping** [freeCodeCamp/freeCodeCamp#53829]
# The Js Curriculumn: It Must Teach _Principles_

## The Problem
I really loved freeCodeCamp's HTML and CSS course. Then I started into their Java Script curriculum and it went really bad. I felt like they wanted me to learn the language by copying, so that when I got to my first independent project I didn't have a clue what to do.

I went and did more research. I found a book that taught me the _principles_ and I did much better. I still really like freeCodeCamp, but this is a BIG problem. Java Script is a fundamental language. If they don't have a good curriculum it may spell bad things for them, and it's certainly pretty hard on students.

## The Fix
freeCodeCamp must teach students the principles behind JS. This will at least double the usefulness of their JS course and will likely increase their popularity as a path to learning to code. Their first challenge should be a whole lot simpler. It should be designed to take the students through the foundation of JS, and not shoot the into the clouds right away. The instructions should have in depth explanations of the reasons why. 

Also, students need to have more time to practice what they learn. The lessons should progress more slowly so that students can retain the information. Right now, each lesson introduces so many new principles that retention is next to impossible

## To Sum Up
freeCodeCamp, as a whole, is a great place and project based learning is the best way to learn. But it must be managed rightly, with reasonable expectations for the students comprehension and retention abilities.

**Comment by rchaurase** [comment]
I believe that the project should be advanced to a certain point, made somewhat challenging, and allowed the student to consider what level is being taught to him.

**Comment by FrontEndWebDeveloping** [comment]
rchaurase, would you expound on what you just said, and clarify? Thankyou. (-:.

**Comment by jdwilkin4** [comment]
> Also, students need to have more time to practice what they learn. The lessons should progress more slowly so that students can retain the information. Right now, each lesson introduces so many new principles that retention is next to impossible

There will be a new project that goes before the RPG project to focus just on JS fundamentals.
The RPG project is also undergoing a rewrite and provide more practice on concepts like you mentioned

**Comment by Supravisor** [comment]
As the course is project based, maybe include links to articles on the freeCodeCamp web site so the learner can easily jump to resources as needed.

**Comment by jdwilkin4** [comment]
@Supravisor 

> maybe include links to articles on the freeCodeCamp web site so the learner can easily jump to resources as needed.

That does sound like a nice idea. 
We would have to see how that works with the translations though.

So that might be something nice to look into for the future changes.

**Comment by Steven-EB** [comment]
I agree, @FrontEndWebDeveloping... I think something that makes the project-based curriculum more challenging is the lack of ability to write lasting comments, since the project code files refresh/load a new version with the progression of each step. This is the 3rd time I’ve done an intro to JS course (Codecademy, legacy fCC, and CS50), and I’m able to muddle my way through it with help from the Bad Website Club bootcamp leads, and my own private mentor, but I find it really hard to keep in mind what each line of code is doing when it gets more complicated. So I’ve shifted to copying the code into my local IDE where I can write comments that don’t get erased.

Perhaps this is actually a good opportunity for the fCC curriculum to model how to write useful comments, as @FrontEndWebDeveloping mentions above – learners could be given an opportunity to write their own comments in the code, and compare them to an experienced engineer’s example of writing comments, which could load with each new step. Maybe even having a separate file, or a section at the top of the JS file, of pseudocode to refer to throughout the project would be a good way for learners to orient themselves to what each line is doing.

**Comment by FrontEndWebDeveloping** [comment]
Thanks @Steven-EB , you have stuck with this "change the Java Script course" situation from almost the beginning. Keep it up.

**Comment by FrontEndWebDeveloping** [comment]
I close this issue, wishing @jdwilkin4 and all those helping her success in creating the new JS curriculum. Thankyou @Steven-EB for sticking in the saddle with this situation. Also I thank any other members of the Bad Coding club who were involved.

I also want to thank @lasjorg, the nicest moderator I've run into on the FCC forum. With out their encouragement, this might not have come to pass.

Thank you all from @FrontEndWebDeveloping .

If any of you wish to contact me you can email leeveyben@gmail.com.
